### PR TITLE
[FEATURE] Use the Lucene Distance Calculation Function in Script Scoring for doing exact search

### LINKED ISSUE #1032: [FEATURE] Use the Lucene Distance Calculation Function in Script Scoring for doing exact search
**Is your feature request related to a problem?**
No.

**What solution would you like?**
Currently in K-NN plugin we are using the custom functions to do the [distance calculation](https://github.com/opensearch-project/k-NN/blob/581faddf10792f7c19b7f4bb9f9deb6f8e77de63/src/main/java/org/opensearch/knn/plugin/script/KNNScoringUtil.java#L151-L167), with Lucene coming up with the support of SIMD we should use Lucene functions(https://github.com/apache/lucene/blob/branch_9_7/lucene/core/src/java/org/apache/lucene/util/VectorUtil.java).

**What alternatives have you considered?**
NA

**Do you have any additional context?**
NA

### PR DESCRIPTION
**Pull request by TrungBui59**
### Description
Change our implementation to use Lucene for calculating distance
 
### Issues Resolved
#1032
 
### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [x] Commits are signed as per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/k-NN/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by navneet1v** [comment]
@TrungBui59 Can you add the entry for this in Changelog.md file

File: src/main/java/org/opensearch/knn/plugin/script/KNNScoringUtil.java [hunk 1]
```diff
@@ -294,11 +295,7 @@ public static float lInfNorm(List<Number> queryVector, KNNVectorScriptDocValues
      */
     public static float innerProduct(float[] queryVector, float[] inputVector) {
         requireEqualDimension(queryVector, inputVector);
```
Comment (navneet1v) [inline src/main/java/org/opensearch/knn/plugin/script/KNNScoringUtil.java:297]: I think this check is present in Lucene code, we can remove it just like we removed from l2Squared

**Review by navneet1v — COMMENTED** [review]
Overall code looks fine to me. @jmazanec15 as you are aware more around the distances on both lucene and k-NN side. Can you please review and see if things are intact.

File: src/main/java/org/opensearch/knn/plugin/script/KNNScoringUtil.java (hunk 1 above)
Comment (TrungBui59) [reply src/main/java/org/opensearch/knn/plugin/script/KNNScoringUtil.java:297]: Sure, I will remove it, also how do I add this to changelog.md. I have read the CONTRIBUTING.md but am still quite confused about how to do it?

File: src/main/java/org/opensearch/knn/plugin/script/KNNScoringUtil.java [hunk 2]
```diff
@@ -150,20 +142,28 @@ public static float cosineSimilarity(List<Number> queryVector, KNNVectorScriptDo
...
+            }
         }
-        float normalizedProduct = normQueryVector * normInputVector;
-        if (normalizedProduct == 0) {
+        if (numZeroInInput == inputVector.length || numZeroInQuery == queryVector.length) {
+            return cosine;
+        }
+        try {
+            cosine = VectorUtil.cosine(queryVector, inputVector);
+        } catch (Exception e) {
```
Comment (jmazanec15) [inline src/main/java/org/opensearch/knn/plugin/script/KNNScoringUtil.java:162]: Can we avoid catching all exceptions here or catch a more specific exception?

File: src/main/java/org/opensearch/knn/plugin/script/KNNScoringUtil.java [hunk 3]
```diff
@@ -48,13 +49,7 @@ private static void requireEqualDimension(final float[] queryVector, final float
...
      */
     public static float l2Squared(float[] queryVector, float[] inputVector) {
-        requireEqualDimension(queryVector, inputVector);
-        float squaredDistance = 0;
-        for (int i = 0; i < inputVector.length; i++) {
-            float diff = queryVector[i] - inputVector[i];
-            squaredDistance += diff * diff;
-        }
-        return squaredDistance;
+        return VectorUtil.squareDistance(queryVector, inputVector);
```
Comment (jmazanec15) [inline src/main/java/org/opensearch/knn/plugin/script/KNNScoringUtil.java:52]: should we requireEqualDim here too?

Comment (TrungBui59) [reply src/main/java/org/opensearch/knn/plugin/script/KNNScoringUtil.java:52]: @jmazanec15 Lucene already have the support for it

File: src/main/java/org/opensearch/knn/plugin/script/KNNScoringUtil.java (hunk 2 above)
Comment (TrungBui59) [reply src/main/java/org/opensearch/knn/plugin/script/KNNScoringUtil.java:162]: @jmazanec15 Sure I will change it

**Comment by navneet1v** [comment]
@TrungBui59 can you resolve the comments so that we can merge the changes

**Comment by navneet1v** [comment]
@TrungBui59 are you still working on this PR?

**Comment by navneet1v** [comment]
@TrungBui59 are you still working on this? if not can we close this PR or someone else can pick up the github issue?

**Comment by aalbahem** [comment]
@navneet1v I would love to pick up this.

**Comment by navneet1v** [comment]
@aalbahem thanks for showing the interest. I can see a PR is already raised for the same: https://github.com/opensearch-project/k-NN/pull/1699.

Let me know if you would be interested in picking up something else.

**Comment by navneet1v** [comment]
Closing this PR due to no further activity from author. Also there is a PR raised for similar feature. Ref: https://github.com/opensearch-project/k-NN/pull/1699

File: src/main/java/org/opensearch/knn/plugin/script/KNNScoringUtil.java (hunk 1 above)
Comment (aalbahem) [reply src/main/java/org/opensearch/knn/plugin/script/KNNScoringUtil.java]: @navneet1v Yes, I would love to start with something small.

**Pull request closed without merging**